SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -0,0 +1,283 @@
+Division by zero is undefined behavior.
+Overflow also leads to undefined behavior; this is a rare case, but can occur, for example, by doing a 32-bit division of -2147483648 by -1.
+-/
+def sdiv? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  if y.toInt = 0
+  then none
+  else
+    let div := (x.toInt / y.toInt)
+    if div < Int.ofNat (2^w)
+      then some $ BitVec.ofInt w div
+      else none
+
+/--
+This instruction returns the unsigned integer remainder of a division. This instruction always performs an unsigned division to get the remainder.
+Note that unsigned integer remainder and signed integer remainder are distinct operations; for signed integer remainder, use ‘srem’.
+Taking the remainder of a division by zero is undefined behavior.
+-/
+def urem? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  if y.toNat = 0
+  then none
+  else some $ BitVec.ofNat w (x.toNat % y.toNat)
+
+def _root_.Int.rem (x y : Int) : Int :=
+  if x ≥ 0 then (x % y) else ((x % y) - y.natAbs)
+
+-- TODO: prove this to make sure it's the right implementation!
+theorem _root_.Int.rem_sign_dividend :
+  ∀ x y, Int.rem x y < 0 ↔ x < 0 :=  by sorry
+
+/--
+This instruction returns the remainder of a division (where the result is either zero or has the same sign as the dividend, op1),
+not the modulo operator (where the result is either zero or has the same sign as the divisor, op2) of a value.
+For more information about the difference, see The Math Forum.
+For a table of how this is implemented in various languages, please see Wikipedia: modulo operation.
+Note that signed integer remainder and unsigned integer remainder are distinct operations; for unsigned integer remainder, use ‘urem’.
+Taking the remainder of a division by zero is undefined behavior.
+For vectors, if any element of the divisor is zero, the operation has undefined behavior.
+Overflow also leads to undefined behavior; this is a rare case, but can occur, for example,
+by taking the remainder of a 32-bit division of -2147483648 by -1.
+(The remainder doesn’t actually overflow, but this rule lets srem be implemented using instructions that return both the result
+of the division and the remainder.)
+-/
+def srem? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  if y.toInt = 0
+  then none
+  else
+    let div := (x.toInt / y.toInt)
+    if div < Int.ofNat (2^w)
+      then some $ BitVec.ofInt w (x.toInt.rem y.toInt)
+      else none
+
+def sshr (a : BitVec n) (s : Nat) := BitVec.sshiftRight a s
+
+/--
+Shift left instruction.
+The value produced is op1 * 2op2 mod 2n, where n is the width of the result.
+If op2 is (statically or dynamically) equal to or larger than the number of
+bits in op1, this instruction returns a poison value.
+-/
+def shl? {m n k} (op1 : BitVec n) (op2 : BitVec m) : Option (BitVec k) :=
+  let bits := op2.toNat -- should this be toInt?
+  if bits >= n then .none
```

Comment:
shouldn't this be `bits >= 2^n` ?

Reply:
I believe `n` is right here, `2^n` is the maximal number that is representable by `n` bits, but since this is a shift, it corresponds to a multiplication by `2^op`, so we do want the logarithm of that multiplication, bits, to be less than the number of bits in `op1`. The power was lost through the formatting in the copy-paste, I fixed the docstring.

Reply:
I believe the explanation now, thanks!